autorest/utility_test.go

```diff
@@ -24,6 +27,64 @@ const (
 	</Person>`
 )
 
+func TestNewDecoderCreatesJSONDecoder(t *testing.T) {
+	d := NewDecoder(EncodedAsJSON, strings.NewReader(jsonT))
+	_, ok := d.(*json.Decoder)
+	if d == nil || !ok {
+		t.Error("autorest: NewDecoder failed to create a JSON decoder when requested")
+	}
+}
+
+func TestNewDecoderCreatesXMLDecoder(t *testing.T) {
+	d := NewDecoder(EncodedAsXML, strings.NewReader(xmlT))
+	_, ok := d.(*xml.Decoder)
+	if d == nil || !ok {
+		t.Error("autorest: NewDecoder failed to create an XML decoder when requested")
+	}
+}
+
+func TestNewDecoderReturnsNilForUnknownEncoding(t *testing.T) {
+	d := NewDecoder("unknown", strings.NewReader(xmlT))
+	if d != nil {
+		t.Error("autorest: NewDecoder created a decoder for an unknown encoding")
+	}
+}
+
+func TestCopyAndDecodeDecodesJSON(t *testing.T) {
+	_, err := CopyAndDecode(EncodedAsJSON, strings.NewReader(jsonT), &mocks.T{})
+	if err != nil {
+		t.Errorf("autorest: CopyAndDecode returned an error with valid JSON - %v", err)
+	}
+}
+
+func TestCopyAndDecodeDecodesXML(t *testing.T) {
+	_, err := CopyAndDecode(EncodedAsXML, strings.NewReader(xmlT), &mocks.T{})
+	if err != nil {
+		t.Errorf("autorest: CopyAndDecode returned an error with valid XML - %v", err)
+	}
+}
+
+func TestCopyAndDecodeReturnsJSONDecodingErrors(t *testing.T) {
+	_, err := CopyAndDecode(EncodedAsJSON, strings.NewReader(jsonT[0:len(jsonT)-2]), &mocks.T{})
+	if err == nil {
+		t.Errorf("autorest: CopyAndDecode failed to return an error with invalid JSON")
+	}
+}
+
+func TestCopyAndDecodeReturnsXMLDecodingErrors(t *testing.T) {
```

Comment:
@ahmetalpbalkan I believe in testing each logical case even if there is overlap. That prevents gaps in the future which simple code coverage analysis will not catch. It's cheap insurance.